nbody_gravitational/NBodyGravitational.jl

```diff
@@ -0,0 +1,173 @@
+module NBodyGravitational
+
+using DiffEqBase, DifferentialEquations, StaticArrays, Plots, StaticArrays
+
+
+grav_const = 6.673e-11
+
+#=
+    Represents a body/particle in an N-body gravitational problem
+    m - mass
+    r - initial position
+    v - initial velocity
+=#
+struct MassBody{mType, cType}
+    m :: mType
+    r :: Vector{cType}
+    v :: Vector{cType}
+    
+    function MassBody{mType,cType}(m::mType, r::Vector{cType}, v::Vector{cType})
+        @assert length(r)==length(v) "The length of the position (r) and velocity (v) components should match"
+        new(m,r,v)
+    end
+end
+
+MassBody(m::mType, r :: Vector{cType}, v :: Vector{cType}) where {mType <: AbstractFloat, cType  <: AbstractFloat} = MassBody{mType, cType}(m,r,v)
+
+#=
+    Contains the necessary information for solving an N-body gravitational problem
+    bodies - interacting bodies/particles
+    tSpan - time, for which we want to evaluate the solution
+    G - gravitational constant
+=#
+struct NBodyGravProblem
+    bodies :: Array{MassBody, 1}
+    tSpan :: Tuple{Float64, Float64}
+    G :: Float64
+end
+
+NBodyGravProblem(bodies, tSpan::Tuple{Float64, Float64}) = NBodyGravProblem(bodies, tSpan, grav_const)
+
+#= 
+    As a quick approach to solving an N-body problem we will just 
+    transform NBodyGravProblem to an ODEProblem
+
+    Here u will be an array consisting of a consecutive positional coordinates r 
+    following by components of velocities v
+=#
+import Base.convert
+convert(::Type{ODEProblem}, x:: NBodyGravProblem) =
+begin
+    n = length(x.bodies)
+    u0 = zeros(6*n);
+    m = zeros(n)
+    
+    for i=1:n
+        ith = 3(i-1)+1;        
+        u0[ith:ith+2] = x.bodies[i].r;
+        u0[3n+ith:3n+ith+2] = x.bodies[i].v 
+        m[i] = x.bodies[i].m
+    end    
+    
+    function ode_system!(du, u, p, t)
+        du[1:3n] = @view u[3n+1:6n];
+        for i=1:n
+            ith = 3(i-1)+1;
+            ri = @SVector [u[ith] , u[ith+1], u[ith+2]]
+            du[3n+ith:3n+ith+2] = 
+            begin
+                accel = @SVector [0.0, 0.0, 0.0]; 
+                for j=1:n
+                    if j!=i
+                        jth = 3(j-1)+1;
+                        rj = @SVector [u[jth], u[jth+1], u[jth+2]]
+                        accel -= x.G*m[j]*(ri-rj)/norm(ri-rj)^3
+                    end
+                end
+                accel
+            end
+        end 
+    end
+
+    ODEProblem(ode_system!, u0, x.tSpan)
+end
+
+import DiffEqBase.solve
+solve(x::NBodyGravProblem; kwargs...) = 
+begin
+    return DiffEqBase.solve(convert(ODEProblem,x); alg_hints=[:stiff], kwargs...)
```

Comment:
Is it usually stiff?

Reply:
I suppose so: if objects move quite close to each other, the step of integration should decrease dramatically. It is a general problem of potentials with singularity in denominator (as Coulomb's one). Plus, real-life number of masses and gravitational constant can also bring errors.  So, by default, I reckon this problem is stiff.

Reply:
I would test it before making that assumption. N-body problems like the Pleiades problem are classic non-stiff tests for ODE integrators.

http://nbviewer.jupyter.org/github/JuliaDiffEq/DiffEqBenchmarks.jl/blob/master/NonStiffODE/Pleiades%20Work-Precision%20Diagrams.ipynb

Also, many symplectic RK integrators were created for these problems, and they are for non-stiff problems. I bet you can find examples of stiff N-body problems, but I would think it's rare given the benchmarks people use. Of course, this is a good reason just just splat the arguments.